File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/statistikk/sykefraværshistorikk/SykefraværForEttKvartalRepositoryJdbcTest.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk;

import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.TestUtils.SISTE_PUBLISERTE_KVARTAL;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.TestUtils.opprettStatistikkForLand;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.TestUtils.slettAllStatistikkFraDatabase;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Varighetskategori._1_DAG_TIL_7_DAGER;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Varighetskategori._8_UKER_TIL_20_UKER;
import static org.assertj.core.api.Java6Assertions.assertThat;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.AppConfigForJdbcTesterConfig;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.Næring;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.Næringskode5Siffer;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.Orgnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.Sektor;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.UnderenhetLegacy;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.bransjeprogram.ArbeidsmiljøportalenBransje;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.bransjeprogram.Bransje;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.ÅrstallOgKvartal;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Varighetskategori;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.kvartalsvis.KvartalsvisSykefraværRepository;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.data.jdbc.DataJdbcTest;
import org.springframework.boot.test.autoconfigure.jdbc.TestDatabaseAutoConfiguration;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@ActiveProfiles("db-test")
@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = {AppConfigForJdbcTesterConfig.class})
@DataJdbcTest(excludeAutoConfiguration = {TestDatabaseAutoConfiguration.class})
public class SykefraværForEttKvartalRepositoryJdbcTest {
  @Autowired private NamedParameterJdbcTemplate jdbcTemplate;

  private KvartalsvisSykefraværRepository kvartalsvisSykefraværprosentRepository;

  @BeforeEach
  public void setUp() {
    kvartalsvisSykefraværprosentRepository = new KvartalsvisSykefraværRepository(jdbcTemplate);
    slettAllStatistikkFraDatabase(jdbcTemplate);
  }

  @AfterEach
  public void tearDown() {
    slettAllStatistikkFraDatabase(jdbcTemplate);
  }

  @Test
  public void hentSykefraværprosentLand__skal_returnere_riktig_sykefravær() {
    opprettStatistikkForLand(jdbcTemplate);
    List<SykefraværForEttKvartal> resultat =
        kvartalsvisSykefraværprosentRepository.hentKvartalsvisSykefraværprosentLand();
    assertThat(resultat.size()).isEqualTo(3);
    assertThat(resultat.get(0))
        .isEqualTo(
            new SykefraværForEttKvartal(
                SISTE_PUBLISERTE_KVARTAL.minusKvartaler(2),
                new BigDecimal(6),
                new BigDecimal(100),
                10));
  }

  @Test
  public void hentSykefraværprosentSektor__skal_returnere_riktig_sykefravær() {
    Sektor statligForvaltning = new Sektor("1", "Statlig forvaltning");
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_sektor (sektor_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:sektor_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(statligForvaltning, 2019, 2, 10, 2, 100));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_sektor (sektor_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:sektor_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(statligForvaltning, 2019, 1, 10, 3, 100));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_sektor (sektor_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:sektor_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(statligForvaltning, 2018, 4, 10, 4, 100));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_sektor (sektor_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:sektor_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(new Sektor("2", "Kommunal forvaltning"), 2018, 4, 10, 5, 100));

    List<SykefraværForEttKvartal> resultat =
        kvartalsvisSykefraværprosentRepository.hentKvartalsvisSykefraværprosentSektor(
            statligForvaltning);
    assertThat(resultat.size()).isEqualTo(3);
    assertThat(resultat.get(0))
        .isEqualTo(
            new SykefraværForEttKvartal(
                new ÅrstallOgKvartal(2018, 4), new BigDecimal(4), new BigDecimal(100), 10));
  }

  @Test
  public void hentSykefraværprosentNæring__skal_returnere_riktig_sykefravær() {
    Næring produksjonAvKlær = new Næring("14", "Produksjon av klær");
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_naring (naring_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:naring_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(produksjonAvKlær, 2019, 2, 10, 2, 100));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_naring (naring_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:naring_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(produksjonAvKlær, 2019, 1, 10, 3, 100));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_naring (naring_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:naring_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(new Næring("85", "Undervisning"), 2018, 4, 10, 5, 100));

    List<SykefraværForEttKvartal> resultat =
        kvartalsvisSykefraværprosentRepository.hentKvartalsvisSykefraværprosentNæring(
            produksjonAvKlær);
    assertThat(resultat.size()).isEqualTo(2);
    assertThat(resultat.get(0))
        .isEqualTo(
            new SykefraværForEttKvartal(
                new ÅrstallOgKvartal(2019, 1), new BigDecimal(3), new BigDecimal(100), 10));
  }

  @Test
  public void hentSykefraværprosentBransje__skal_returnere_riktig_sykefravær() {
    Næring produksjonAvKlær = new Næring("14", "Produksjon av klær");
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_naring5siffer "
            + "(naring_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:naring_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(new Næring("87101", "Somatiske spesialsykehjem"), 2019, 2, 10, 5, 100));

    jdbcTemplate.update(
        "insert into sykefravar_statistikk_naring5siffer "
            + "(naring_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:naring_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(new Næring("87101", "Somatiske spesialsykehjem"), 2019, 1, 10, 1, 100));

    jdbcTemplate.update(
        "insert into sykefravar_statistikk_naring5siffer "
            + "(naring_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:naring_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(new Næring("87102", "Somatiske sykehjem"), 2019, 1, 10, 7, 100));

    jdbcTemplate.update(
        "insert into sykefravar_statistikk_naring5siffer "
            + "(naring_kode, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:naring_kode, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(new Næring("87301", "Aldershjem"), 2018, 4, 10, 6, 100));

    Bransje sykehjem =
        new Bransje(ArbeidsmiljøportalenBransje.SYKEHUS, "Sykehjem", "87101", "87102");
    List<SykefraværForEttKvartal> resultat =
        kvartalsvisSykefraværprosentRepository.hentKvartalsvisSykefraværprosentBransje(sykehjem);
    assertThat(resultat.size()).isEqualTo(2);
    assertThat(resultat.get(0))
        .isEqualTo(
            new SykefraværForEttKvartal(
                new ÅrstallOgKvartal(2019, 1), new BigDecimal(8), new BigDecimal(200), 20));
  }

  @Test
  public void hentSykefraværprosentVirksomhet__skal_returnere_riktig_sykefravær() {
    UnderenhetLegacy barnehage =
        new UnderenhetLegacy(
            new Orgnr("999999999"),
            new Orgnr("1111111111"),
            "test Barnehage",
            new Næringskode5Siffer("88911", "Barnehage"),
            10);
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_virksomhet (orgnr, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:orgnr, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(barnehage.getOrgnr(), 2019, 2, 10, 2, 100, _1_DAG_TIL_7_DAGER));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_virksomhet (orgnr, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:orgnr, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(new Orgnr("987654321"), 2019, 1, 10, 3, 100, _1_DAG_TIL_7_DAGER));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_virksomhet (orgnr, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:orgnr, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(barnehage.getOrgnr(), 2018, 4, 10, 5, 100, _1_DAG_TIL_7_DAGER));

    List<SykefraværForEttKvartal> resultat =
        kvartalsvisSykefraværprosentRepository.hentKvartalsvisSykefraværprosentVirksomhet(
            barnehage);
    assertThat(resultat.size()).isEqualTo(2);
    assertThat(resultat.get(0))
        .isEqualTo(
            new SykefraværForEttKvartal(
                new ÅrstallOgKvartal(2018, 4), new BigDecimal(5), new BigDecimal(100), 10));
  }

  @Test
  public void hentSykefraværprosentVirksomhet__skal_summere_sykefravær_på_varighet() {
    UnderenhetLegacy barnehage =
        new UnderenhetLegacy(
            new Orgnr("999999999"),
            new Orgnr("1111111111"),
            "test Barnehage",
            new Næringskode5Siffer("88911", "Barnehage"),
            10);
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_virksomhet (orgnr, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk, varighet) "
            + "VALUES (:orgnr, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk, :varighet)",
        parametre(barnehage.getOrgnr(), 2019, 2, 10, 2, 100, _1_DAG_TIL_7_DAGER));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_virksomhet (orgnr, arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk, varighet) "
            + "VALUES (:orgnr, :arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk, :varighet)",
        parametre(barnehage.getOrgnr(), 2019, 2, 10, 5, 100, _8_UKER_TIL_20_UKER));
    List<SykefraværForEttKvartal> resultat =
        kvartalsvisSykefraværprosentRepository.hentKvartalsvisSykefraværprosentVirksomhet(
            barnehage);
    assertThat(resultat)
        .isEqualTo(
            Arrays.asList(
                new SykefraværForEttKvartal(
                    new ÅrstallOgKvartal(2019, 2), new BigDecimal(7), new BigDecimal(200), 20)));
  }

  @Test
  public void hentSykefraværprosentLand__maskerer_sf_dersom_antall_ansatte_er_for_lav() {
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_land (arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(2019, 2, 4, 4, 100));
    jdbcTemplate.update(
        "insert into sykefravar_statistikk_land (arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) "
            + "VALUES (:arstall, :kvartal, :antall_personer, :tapte_dagsverk, :mulige_dagsverk)",
        parametre(2019, 1, 10, 5, 100));

    List<SykefraværForEttKvartal> resultat =
        kvartalsvisSykefraværprosentRepository.hentKvartalsvisSykefraværprosentLand();

    assertThat(resultat.size()).isEqualTo(2);

    SykefraværForEttKvartal ikkeMaskertSykefraværForEttKvartal = resultat.get(0);
    assertThat(ikkeMaskertSykefraværForEttKvartal.isErMaskert()).isFalse();
    assertThat(ikkeMaskertSykefraværForEttKvartal.getProsent().setScale(2))
        .isEqualTo(new BigDecimal(5).setScale(2));

    SykefraværForEttKvartal maskertSykefraværForEttKvartal = resultat.get(1);
    assertThat(maskertSykefraværForEttKvartal.isErMaskert()).isTrue();
    assertThat(maskertSykefraværForEttKvartal.getProsent()).isNull();
  }

  private MapSqlParameterSource parametre(
      int årstall, int kvartal, int antallPersoner, int tapteDagsverk, int muligeDagsverk) {
    return new MapSqlParameterSource()
        .addValue("arstall", årstall)
        .addValue("kvartal", kvartal)
        .addValue("antall_personer", antallPersoner)
        .addValue("tapte_dagsverk", tapteDagsverk)
        .addValue("mulige_dagsverk", muligeDagsverk);
  }

  private MapSqlParameterSource parametre(
      Sektor sektor,
      int årstall,
      int kvartal,
      int antallPersoner,
      int tapteDagsverk,
      int muligeDagsverk) {
    return parametre(årstall, kvartal, antallPersoner, tapteDagsverk, muligeDagsverk)
        .addValue("sektor_kode", sektor.getKode());
  }

  private MapSqlParameterSource parametre(
      Næring næring,
      int årstall,
      int kvartal,
      int antallPersoner,
      int tapteDagsverk,
      int muligeDagsverk) {
    return parametre(årstall, kvartal, antallPersoner, tapteDagsverk, muligeDagsverk)
        .addValue("naring_kode", næring.getKode());
  }

  private MapSqlParameterSource parametre(
      Orgnr orgnr,
      int årstall,
      int kvartal,
      int antallPersoner,
      int tapteDagsverk,
      int muligeDagsverk,
      Varighetskategori varighet) {
    return parametre(årstall, kvartal, antallPersoner, tapteDagsverk, muligeDagsverk)
        .addValue("orgnr", orgnr.getVerdi())
        .addValue("varighet", varighet.kode);
  }
}

File: src/test/kotlin/no/nav/arbeidsgiver/sykefravarsstatistikk/api/infrastruktur/database/SykefraværForEttKvartalRepositoryJdbcTest.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.infrastruktur.database

import config.AppConfigForJdbcTesterConfig
import ia.felles.definisjoner.bransjer.BransjeId
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.aggregertOgKvartalsvisSykefraværsstatistikk.domene.Varighetskategori
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.fellesdomene.*
import org.assertj.core.api.AssertionsForClassTypes
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.data.jdbc.DataJdbcTest
import org.springframework.boot.test.autoconfigure.jdbc.TestDatabaseAutoConfiguration
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.math.BigDecimal
import ia.felles.definisjoner.bransjer.Bransje
import org.jetbrains.exposed.sql.deleteAll

@ActiveProfiles("db-test")
@ExtendWith(SpringExtension::class)
@ContextConfiguration(classes = [AppConfigForJdbcTesterConfig::class])
@DataJdbcTest(excludeAutoConfiguration = [TestDatabaseAutoConfiguration::class])
open class SykefraværForEttKvartalRepositoryJdbcTest {

    @Autowired
    private lateinit var sykefraværStatistikkVirksomhetRepository: SykefravarStatistikkVirksomhetRepository

    @Autowired
    private lateinit var sykefraværStatistikkLandRepository: SykefraværStatistikkLandRepository

    @Autowired
    private lateinit var sykefraværStatistikkSektorRepository: SykefraværStatistikkSektorRepository

    @Autowired
    private lateinit var sykefraværStatistikkNæringRepository: SykefraværStatistikkNæringRepository

    @Autowired
    private lateinit var sykefraværStatistikkNæringskodeRepository: SykefraværStatistikkNæringskodeRepository

    val SISTE_PUBLISERTE_KVARTAL = ÅrstallOgKvartal(2022, 1)

    @BeforeEach
    fun setUp() {
        with(sykefraværStatistikkVirksomhetRepository) { transaction { deleteAll() } }
        with(sykefraværStatistikkLandRepository) { transaction { deleteAll() } }
    }

    @AfterEach
    fun tearDown() {
        with(sykefraværStatistikkVirksomhetRepository) { transaction { deleteAll() } }
        with(sykefraværStatistikkLandRepository) { transaction { deleteAll() } }
    }

    @Test
    fun hentSykefraværprosentLand__skal_returnere_riktig_sykefravær() {
        sykefraværStatistikkLandRepository.settInn(
            listOf(
                SykefraværsstatistikkLand(
                    årstall = this.SISTE_PUBLISERTE_KVARTAL.årstall,
                    kvartal = this.SISTE_PUBLISERTE_KVARTAL.kvartal,
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal("4.0"),
                    muligeDagsverk = BigDecimal("100.0")
                ),
                SykefraværsstatistikkLand(
                    årstall = this.SISTE_PUBLISERTE_KVARTAL.minusKvartaler(1).årstall,
                    kvartal = this.SISTE_PUBLISERTE_KVARTAL.minusKvartaler(1).kvartal,
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal("5.0"),
                    muligeDagsverk = BigDecimal("100.0")
                ),
                SykefraværsstatistikkLand(
                    årstall = this.SISTE_PUBLISERTE_KVARTAL.minusKvartaler(2).årstall,
                    kvartal = this.SISTE_PUBLISERTE_KVARTAL.minusKvartaler(2).kvartal,
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal("6.0"),
                    muligeDagsverk = BigDecimal("100.0")
                ),
            )
        )
        val resultat = sykefraværStatistikkLandRepository.hentAlt()
        AssertionsForClassTypes.assertThat(resultat.size).isEqualTo(3)
        AssertionsForClassTypes.assertThat(resultat[0])
            .isEqualTo(
                SykefraværForEttKvartal(
                    SISTE_PUBLISERTE_KVARTAL.minusKvartaler(2),
                    BigDecimal(6),
                    BigDecimal(100),
                    10
                )
            )
    }

    @Test
    fun hentSykefraværprosentSektor__skal_returnere_riktig_sykefravær() {

        sykefraværStatistikkSektorRepository.settInn(
            listOf(
                SykefraværsstatistikkSektor(
                    årstall = 2019,
                    kvartal = 2,
                    sektor = Sektor.STATLIG.sektorkode,
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal(2),
                    muligeDagsverk = BigDecimal(100)
                ),
                SykefraværsstatistikkSektor(
                    årstall = 2019,
                    kvartal = 1,
                    sektor = Sektor.STATLIG.sektorkode,
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal(2),
                    muligeDagsverk = BigDecimal(100)
                ),
                SykefraværsstatistikkSektor(
                    årstall = 2018,
                    kvartal = 4,
                    sektor = Sektor.STATLIG.sektorkode,
                    antallPersoner = 10,
                    tapteDagsverk = 4.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
                SykefraværsstatistikkSektor(
                    årstall = 2018,
                    kvartal = 4,
                    sektor = Sektor.KOMMUNAL.sektorkode,
                    antallPersoner = 10,
                    tapteDagsverk = 5.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),

                )
        )

        val resultat =
            sykefraværStatistikkSektorRepository.hentKvartalsvisSykefraværprosent(
                Sektor.STATLIG
            )
        AssertionsForClassTypes.assertThat(resultat.size).isEqualTo(3)
        AssertionsForClassTypes.assertThat(resultat[0])
            .isEqualTo(
                SykefraværForEttKvartal(
                    ÅrstallOgKvartal(2018, 4), BigDecimal(4), BigDecimal(100), 10
                )
            )
    }

    @Test
    fun hentSykefraværprosentNæring__skal_returnere_riktig_sykefravær() {
        val produksjonAvKlær = Næring("14")
        sykefraværStatistikkNæringRepository.settInn(
            listOf(
                SykefraværsstatistikkForNæring(
                    næring = produksjonAvKlær.tosifferIdentifikator,
                    årstall = 2019,
                    kvartal = 2,
                    antallPersoner = 10,
                    tapteDagsverk = 2.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
                SykefraværsstatistikkForNæring(
                    næring = produksjonAvKlær.tosifferIdentifikator,
                    årstall = 2019,
                    kvartal = 1,
                    antallPersoner = 10,
                    tapteDagsverk = 3.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
                SykefraværsstatistikkForNæring(
                    næring = Næring("85").tosifferIdentifikator,
                    årstall = 2018,
                    kvartal = 4,
                    antallPersoner = 10,
                    tapteDagsverk = 5.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
            )
        )
        val resultat = sykefraværStatistikkNæringRepository.hentKvartalsvisSykefraværprosent(produksjonAvKlær)
        AssertionsForClassTypes.assertThat(resultat.size).isEqualTo(2)
        AssertionsForClassTypes.assertThat(resultat[0])
            .isEqualTo(
                SykefraværForEttKvartal(
                    ÅrstallOgKvartal(2019, 1), BigDecimal(3), BigDecimal(100), 10
                )
            )
    }

    @Test
    fun `hentKvartalsvisSykefraværprosent skal returnere riktig sykefravær`() {
        sykefraværStatistikkNæringskodeRepository.settInn(
            listOf(
                SykefraværsstatistikkForNæringskode(
                    næringskode = Næringskode("87101").femsifferIdentifikator,
                    årstall = 2019,
                    kvartal = 2,
                    antallPersoner = 10,
                    tapteDagsverk = 5.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
                SykefraværsstatistikkForNæringskode(
                    næringskode = Næringskode("87101").femsifferIdentifikator,
                    årstall = 2019,
                    kvartal = 1,
                    antallPersoner = 10,
                    tapteDagsverk = 1.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
                SykefraværsstatistikkForNæringskode(
                    næringskode = Næringskode("87102").femsifferIdentifikator,
                    årstall = 2019,
                    kvartal = 1,
                    antallPersoner = 10,
                    tapteDagsverk = 7.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
                SykefraværsstatistikkForNæringskode(
                    næringskode = Næringskode("87301").femsifferIdentifikator,
                    årstall = 2018,
                    kvartal = 4,
                    antallPersoner = 10,
                    tapteDagsverk = 6.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal()
                ),
            )
        )

        val resultat =
            sykefraværStatistikkNæringskodeRepository.hentKvartalsvisSykefraværprosent(
                (Bransje.SYKEHJEM.bransjeId as BransjeId.Næringskoder).næringskoder.map { Næringskode(it) },
            )
        AssertionsForClassTypes.assertThat(resultat.size).isEqualTo(2)
        AssertionsForClassTypes.assertThat(resultat[0])
            .isEqualTo(
                SykefraværForEttKvartal(
                    ÅrstallOgKvartal(2019, 1), BigDecimal(8), BigDecimal(200), 20
                )
            )
    }

    @Test
    fun hentSykefraværprosentVirksomhet__skal_returnere_riktig_sykefravær() {
        val barnehage = Underenhet.Næringsdrivende(
            Orgnr("999999999"),
            Orgnr("1111111111"),
            "test Barnehage",
            Næringskode("88911"),
            10
        )

        sykefraværStatistikkVirksomhetRepository.settInn(
            listOf(
                SykefraværsstatistikkVirksomhet(
                    årstall = 2019,
                    kvartal = 2,
                    orgnr = barnehage.orgnr.verdi,
                    varighet = Varighetskategori._1_DAG_TIL_7_DAGER.kode,
                    rectype = "2",
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal(2),
                    muligeDagsverk = BigDecimal(100)
                ),
                SykefraværsstatistikkVirksomhet(
                    årstall = 2019,
                    kvartal = 1,
                    orgnr = "987654321",
                    varighet = Varighetskategori._1_DAG_TIL_7_DAGER.kode,
                    rectype = "2",
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal(3),
                    muligeDagsverk = BigDecimal(100)
                ),
                SykefraværsstatistikkVirksomhet(
                    årstall = 2018,
                    kvartal = 4,
                    antallPersoner = 10,
                    tapteDagsverk = 5.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal(),
                    orgnr = barnehage.orgnr.verdi,
                    varighet = Varighetskategori._1_DAG_TIL_7_DAGER.kode,
                    rectype = "2"
                )
            )
        )

        val resultat = sykefraværStatistikkVirksomhetRepository.hentAlt(barnehage.orgnr)

        AssertionsForClassTypes.assertThat(resultat.size).isEqualTo(2)
        AssertionsForClassTypes.assertThat(resultat[0])
            .isEqualTo(
                SykefraværForEttKvartal(
                    ÅrstallOgKvartal(2018, 4), BigDecimal(5), BigDecimal(100), 10
                )
            )
    }

    @Test
    fun hentSykefraværprosentVirksomhet__skal_summere_sykefravær_på_varighet() {
        val barnehage = Underenhet.Næringsdrivende(
            Orgnr("999999999"),
            Orgnr("1111111111"),
            "test Barnehage",
            Næringskode("88911"),
            10
        )

        sykefraværStatistikkVirksomhetRepository.settInn(
            listOf(
                SykefraværsstatistikkVirksomhet(
                    årstall = 2019,
                    kvartal = 2,
                    orgnr = barnehage.orgnr.verdi,
                    varighet = Varighetskategori._1_DAG_TIL_7_DAGER.kode,
                    rectype = "2",
                    antallPersoner = 10,
                    tapteDagsverk = BigDecimal(2),
                    muligeDagsverk = BigDecimal(100)
                ),
                SykefraværsstatistikkVirksomhet(
                    årstall = 2019,
                    kvartal = 2,
                    antallPersoner = 10,
                    tapteDagsverk = 5.toBigDecimal(),
                    muligeDagsverk = 100.toBigDecimal(),
                    orgnr = barnehage.orgnr.verdi,
                    varighet = Varighetskategori._8_UKER_TIL_20_UKER.kode,
                    rectype = "2"
                )
            )
        )

        val resultat = sykefraværStatistikkVirksomhetRepository.hentAlt(barnehage.orgnr)

        AssertionsForClassTypes.assertThat(resultat)
            .isEqualTo(
                listOf(
                    SykefraværForEttKvartal(
                        ÅrstallOgKvartal(2019, 2), BigDecimal(7), BigDecimal(200), 20
                    )
                )
            )
    }

}
